Dodano encje rentalItem

ZuberRS03 opened the pull request:
Zgadnie z poradnikiem dodałem encje rentalItem.  
Encja zawiera kolumny: id(Long), name(String), description(String), prise(Float), quantity(int).

Do wytestowania:
`POST` z parametrem `Body` jako JSON (http://localhost:8080/api/v1/rentalItem/add) 
Dodaje przedmiot o podanych parametrach.
przykładowa zawartość w JSON: 
```json
{
  "name": "czekan",
  "description": "jakiś piękny opis",
  "price": 36.7,
  "quantity": 4
}
```

`GET `
http://localhost:8080/api/v1/rentalItem - daje wszystkie przedmioty
http://localhost:8080/api/v1/rentalItem/* - daje przedmiot o konkretnym `ID`

`PUT` z `ID` przedmiotu do zmiany w linku, a parametrami do zmiany w `Body`
Zmieniana jest dowolna ilość parametrów w przedmiocie o podanym `ID`
http://localhost:8080/api/v1/rentalItem/1
```json
{
  "name": "nowy czekan",
  "description": "jakiś jeszcze piękniejszy opis",
  "price": 34.7,
  "quantity": 6
}
```

`DELETE` z parametrem w linku
usuwa przedmiot o podanym `ID`
http://localhost:8080/api/v1/rentalItem/delete/*

ZuberRS03 committed:
rentalItem entiti v1

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemController.java:
```diff
@@ -0,0 +1,48 @@
+        return rentalItemService.getRentalItemById(rentalItemId);
+    }
+
+    @PostMapping
```
Dodałbym tutaj specyficzną ścieżkę dodawania - np. '/add` - żeby to było jakkolwiek oddzielone :)

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PostMapping
+    public void addNewRentalItem(@RequestBody RentalItem rentalItem) {
+        rentalItemService.addNewRentalItem(rentalItem);
```
tutaj też zastanowiłbym się nad jakąkolwiek zwrotką do użytkownika - natomiast to powiem co i jak po moim pull request, gdzie pokaże wam jak takie zwrotki dawać

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PutMapping(path = "{rentalItemId}")
+    public void updateRentalItem(
+            @PathVariable("rentalItemId") Long rentalItemId,
```
No tutaj zdecydowanie zastanowiłbym się nad `@RequestBody` niż nad `@RequestParam`

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemController.java:
```diff
@@ -0,0 +1,48 @@
+        rentalItemService.addNewRentalItem(rentalItem);
+    }
+
+    @DeleteMapping(path = "{rentalItemId}")
```
Tutaj dodałbym ścieżkę `/delete/{rentalItemId}`

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
Jako, że to aplikacja na projekt na uczelnię, dodawajmy od razu wszystkie zwrotki do użytkownika (przynamniej informacje co się dzieje, już te HttpResponse mogą być po angielsku) w języku polskim

xEdziu reviewed (changes requested):
Działać działa - to najważniejsze, zmiany raczej kosmetyczne i jedna użytkowa, mało do poprawy, powinno zająć niecałą chwilę :)

overall, bardzo dobra robota

ZuberRS03 committed:
Pierwsza część poprawek

Na razie dodałem specyficzną ścieżkę dodawania "/add"
Przerobiłem funkcję update żeby przyjmowała body oraz w tej funkcji zrobiłem zwrotki dla użytkownika.
Do zrobienia zostało:
dodanie ścieżki /delete/
reszta zwrotek dla użytkownika

ZuberRS03 committed:
Poprawki do encji RentalItem

Dodałem zwrotki przy różnych operacjach oraz GET działa teraz przez DTO

ZuberRS03 commented:
Jak coś to w pierwszym komentarzu edytowałem tak żeby linki i treść była aktualna

xEdziu committed:
Updated pom.xml to the newest version

xEdziu commented:
![image](https://github.com/user-attachments/assets/9c659f08-0b6f-47de-9e5b-a747f199891d)

Mogę dodać kilkukrotnie ten sam item - czy jest to zamierzone? XD może jakieś sprawdzenie czy np. dany tytuł istnieje czy np. badziej zwiększenie quantity już istniejącego rekordu

xEdziu committed:
Disabled security for a testing purposes

xEdziu commented:
Plus pytanie - czy chcemy zrobić rozdzielenie na akcje typu użytkownik, admin, ogolnodostępne? 
Przydało by się to @ZuberRS03

xEdziu reviewed (changes requested):
Tutaj pom.xml zmieniłem, żeby już było tak jak jest w moim

xEdziu committed:
Merge branch 'main' into encja-rentalItem

ZuberRS03 committed:
Dodano opcja wypożyczenia

Dodałem opcje wypożyczeń wzorując się na sklepie. Do poprawienia są daty wypożyczenia i zwrotu.

ZuberRS03 committed:
Merge pull request #27 from xEdziu/main

aktualizacja

ZuberRS03 committed:
Dorobiłem historie wypożyczeń

Co trzeba dorobić lub poprawić jest zapisane w kodzie jako "TODO"

ZuberRS03 committed:
Merge pull request #32 from xEdziu/main

aktualizacja

ZuberRS03 committed:
Rezerwacje, daty, zwracanie przedmiotu

Dodałem do funkcji wypożyczania zczytywanie dat, jeżeli data wynajęcia będzie inna niż obecna to jest to tratowane jako rezerwacja, taka jak w chwili wynajmowania to jest to od razu wynajęcie, daty są walidowane,
Dodałem funkcję zwrotu przedmiotu. Po zwrocie nie znika ona z logów tylko przechodzi w status RETURNED.

ZuberRS03 committed:
Dodano status każdego przedmiotu

Teraz każdy przedmiot może zostać ukryty lub pokazany użytkownikowi bez usuwania go z bazy danych

ZuberRS03 committed:
konkretny użytkownik, dalsza walidacja daty, DTO dla usera

Teraz już jest wykrywany odpowiedni użytkownik i do niego są przypisywane wszystkie wypożyczenia.
Przy funkcji DeleteRentalItem wyświetla się już odpowiedni błąd, a nie 500-tka.
Dodałem dodatkową walidacje czy data oddania nie jest wcześniejsza niż data wynajmu.
Dodałem DTO żeby user nie widział wszystkich informacji o przedmiocie.

ZuberRS03 commented:
Dobra prawdopodobnie skończyłem z tę encją, potestuj i w razie co będę łatał. Przesyłam Ci wszystkie moje zapytania.
```xml
<folderState>
                <option name="id" value="21412d3e-c668-451f-adaa-c6d4973da586" />
                <option name="name" value="Rental Items" />
                <option name="requests">
                  <list>
                    <requestState>
                      <option name="description" value="Pobierz wszystkie itemy z wypożyczalni" />
                      <option name="id" value="26d233b6-ca7b-4a21-9b8e-8ef14d95bcb2" />
                      <option name="method" value="GET" />
                      <option name="name" value="All item for Admin" />
                      <option name="sortWeight" value="500000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/admin/rentalItem/all" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Pobierz wszyskie aktywne itemy z wypożyczlni" />
                      <option name="id" value="03b92bb6-8168-4dac-8234-c024cd0457ce" />
                      <option name="method" value="GET" />
                      <option name="name" value="All active Items" />
                      <option name="sortWeight" value="1000000" />
                      <option name="url" value="http://localhost:8080/api/v1/rentalItem/all" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Pobierz konkretny item z wypożyczlni" />
                      <option name="id" value="a745e927-a068-4074-8a01-fb5b53124d11" />
                      <option name="method" value="GET" />
                      <option name="name" value="One Item" />
                      <option name="sortWeight" value="2000000" />
                      <option name="url" value="http://localhost:8080/api/v1/rentalItem/1" />
                    </requestState>
                    <requestState>
                      <option name="body">
                        <bodyState>
                          <option name="raw" value="{&#10;  &quot;name&quot;: &quot;zmieniona magnezja&quot;,&#10;  &quot;description&quot;: &quot;opis jakis tam poprawiony&quot;,&#10;  &quot;price&quot;: 16.5,&#10;  &quot;quantity&quot;: 29,&#10;  &quot;status&quot;: &quot;INACTIVE&quot;&#10;}" />
                          <option name="type" value="JSON" />
                        </bodyState>
                      </option>
                      <option name="description" value="Zaaktualizuj item o danym id" />
                      <option name="id" value="fcd2602b-17d5-4bc0-8060-0ea18c23b998" />
                      <option name="method" value="PUT" />
                      <option name="name" value="Update item" />
                      <option name="queryParams">
                        <list>
                          <queryParamState>
                            <option name="disabled" value="true" />
                            <option name="key" value="name" />
                            <option name="value" value="nowyCzekan" />
                          </queryParamState>
                          <queryParamState>
                            <option name="disabled" value="true" />
                            <option name="key" value="description" />
                            <option name="value" value="jeszcze ładniejszy opis" />
                          </queryParamState>
                        </list>
                      </option>
                      <option name="sortWeight" value="3000000" />
                      <option name="url" value="PUT http://localhost:8080/api/v1/authorized/admin/rentalItem/update/1" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Usuń item o danym id" />
                      <option name="id" value="e9cbe742-621f-4ae5-882e-3bf1b82a42eb" />
                      <option name="method" value="DELETE" />
                      <option name="name" value="Deleate item by id" />
                      <option name="sortWeight" value="4000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/admin/rentalItem/delete/1" />
                    </requestState>
                    <requestState>
                      <option name="body">
                        <bodyState>
                          <option name="raw" value="{&#10;  &quot;name&quot;: &quot;worek&quot;,&#10;  &quot;description&quot;: &quot;jakiś piękny opis worka&quot;,&#10;  &quot;price&quot;: 12.3,&#10;  &quot;quantity&quot;: 5,&#10;  &quot;status&quot;: &quot;ACTIVE&quot;,&#10;}" />
                          <option name="type" value="JSON" />
                        </bodyState>
                      </option>
                      <option name="description" value="Dodaj przedmiot do wypożyczlni" />
                      <option name="id" value="e9077354-50d9-46c6-a259-7e28ad9dbb24" />
                      <option name="method" value="POST" />
                      <option name="name" value="Add new item" />
                      <option name="sortWeight" value="5000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/admin/rentalItem/add" />
                    </requestState>
                    <requestState>
                      <option name="body">
                        <bodyState>
                          <option name="raw" value="{&#10;  &quot;rentRequests&quot;: [&#10;    {&#10;      &quot;itemId&quot;: 1,&#10;      &quot;quantity&quot;: 2&#10;    },&#10;    {&#10;      &quot;itemId&quot;: 2,&#10;      &quot;quantity&quot;: 1&#10;    }&#10;  ],&#10;  &quot;rentDate&quot;: &quot;2024-12-01T10:00:00.000Z&quot;,&#10;  &quot;returnDate&quot;: &quot;2025-01-05T10:00:00.000Z&quot;&#10;}" />
                          <option name="type" value="JSON" />
                        </bodyState>
                      </option>
                      <option name="description" value="Wypożycz kilka itemów w kilku sztukach" />
                      <option name="id" value="0ebdd99f-e704-4b7a-8b01-5450f4a2ec59" />
                      <option name="method" value="POST" />
                      <option name="name" value="Rent Many-Many" />
                      <option name="sortWeight" value="6000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/rentalItem/rent" />
                    </requestState>
                    <requestState>
                      <option name="body">
                        <bodyState>
                          <option name="raw" value="{&#10;  &quot;rentRequests&quot;: [&#10;    {&#10;      &quot;itemId&quot;: 1,&#10;      &quot;quantity&quot;: 1&#10;    }&#10;  ],&#10;  &quot;rentDate&quot;: &quot;2024-12-01T10:00:00.000Z&quot;,&#10;  &quot;returnDate&quot;: &quot;2025-01-05T10:00:00.000Z&quot;&#10;}" />
                          <option name="type" value="JSON" />
                        </bodyState>
                      </option>
                      <option name="description" value="Wypożycz jeden item w jednej sztuce" />
                      <option name="id" value="0ceef2df-332e-4f0d-ad55-34702745d86a" />
                      <option name="method" value="POST" />
                      <option name="name" value="Rent 1-1" />
                      <option name="sortWeight" value="7000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/rentalItem/rent" />
                    </requestState>
                    <requestState>
                      <option name="body">
                        <bodyState>
                          <option name="raw" value="{&#10;  &quot;rentRequests&quot;: [&#10;    {&#10;      &quot;itemId&quot;: 1,&#10;      &quot;quantity&quot;: 3&#10;    }&#10;  ],&#10;  &quot;rentDate&quot;: &quot;2024-12-01T10:00:00.000Z&quot;,&#10;  &quot;returnDate&quot;: &quot;2025-01-05T10:00:00.000Z&quot;&#10;}" />
                          <option name="type" value="JSON" />
                        </bodyState>
                      </option>
                      <option name="description" value="Wypożycz jeden item w kilku sztukach" />
                      <option name="id" value="51ff4906-362d-429f-920e-85c72a5ff49d" />
                      <option name="method" value="POST" />
                      <option name="name" value="Rent 1-Many" />
                      <option name="sortWeight" value="8000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/rentalItem/rent" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Pobierz wszystkie wypożyczenia" />
                      <option name="id" value="5d3e990e-42a2-4ce1-a9f9-ee88a8496cb8" />
                      <option name="method" value="GET" />
                      <option name="name" value="All rented" />
                      <option name="sortWeight" value="9000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/admin/rentalItem/history" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Pobierz wszystkie wypożyczenia konkretnego usera" />
                      <option name="id" value="0abff846-d936-4d3a-9168-9643dc20ed72" />
                      <option name="method" value="GET" />
                      <option name="name" value="User rental history" />
                      <option name="sortWeight" value="10000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/rentalItem/rentHistory" />
                    </requestState>
                    <requestState>
                      <option name="description" value="Zwraca przedmioty z konkretnego wypożyczenia" />
                      <option name="id" value="5a16d3e2-50f4-46d5-b36d-d6a21948412a" />
                      <option name="method" value="POST" />
                      <option name="name" value="Return the rental" />
                      <option name="sortWeight" value="11000000" />
                      <option name="url" value="http://localhost:8080/api/v1/authorized/rentalItem/return/4" />
                    </requestState>
                  </list>
                </option>
                <option name="sortWeight" value="3000000" />
              </folderState>
```

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+    public RentalItem getRentalItemById(Long rentalItemId) {
+        boolean exists = rentalItemRepository.existsById(rentalItemId);
+        if( !exists ) {
+            throw new IllegalArgumentException("Przedmiot o id " + rentalItemId + " nie istnieje");
```
![image](https://github.com/user-attachments/assets/7fdc93ae-908a-41b3-9a5b-37e5acf48eb0)


1. Nie `IlleagalArgument` tylko, `IllegalState` - obsługujemy drugi błąd, który zwraca `401`, a nie `500`
2. To chyba podlega pod `ResourceNotFoundException` :D

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+        // Sprawdzenie, czy przedmiot figuruje w liście wypożyczeń
+        boolean isRented = rentRepository.existsByItems_RentalItem_Id(rentalItemId);
+        if (isRented) {
+            throw new IllegalStateException("Przedmiot o id " + rentalItemId + " jest/był wypożyczony i nie może zostać usunięty");
```
![image](https://github.com/user-attachments/assets/cc6ede5a-789f-46ab-9b34-c6abb16a01bf)

Nie robiłbym tego w ten sposób - może po prostu zdeaktywować ten element? żeby tak czy siak był w historii wypożyczeń, ale nie możliwy do dalszego wypożyczenia?

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+    public Map<String, Object> addNewRentalItem(RentalItem rentalItem) {
+        Optional<RentalItem> rentalItemOptional = rentalItemRepository.findRentalItemByName(rentalItem.getName());
+        if (rentalItemOptional.isPresent()) {
+            throw new IllegalArgumentException("Przedmiot o nazwie: " + rentalItem.getName() + " już istnieje");
```
![image](https://github.com/user-attachments/assets/d4f769da-9dc2-40d1-944c-25b1011f3f0a)

Znów `IllegalState` powinien być zamiast `IllegalArgument`

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+    public Map<String, Object> deleteRentalItem(Long rentalItemId) {
+        boolean exists = rentalItemRepository.existsById(rentalItemId);
+        if( !exists ) {
+            throw new IllegalStateException("Przedmiot o id " + rentalItemId + " nie istnieje");
```
![image](https://github.com/user-attachments/assets/9d1d63b4-0266-4d3e-a1de-4688b8ee5057)

a tutaj `ResourceNotFoundException`

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+            throw new IllegalArgumentException("Należy podać jeden z parametrów");
+        }
+        if(!rentalItemRepository.existsById(rentalItemId)) {
+            throw new IllegalArgumentException("Przedmiot o id " + rentalItemId + " nie istnieje");
```
`ResourceNotFoundException` :D

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+
+        if(rentalItem.getPrice() != null) {
+            if(rentalItemToUpdate.getPrice() == null) {
+                throw new IllegalStateException("Cena nie możę być pusta");
```
literówka

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+                rentalItem.getQuantity() == null &&
+                rentalItem.getDescription() == null &&
+                rentalItem.getStatus() == null) {
+            throw new IllegalArgumentException("Należy podać jeden z parametrów");
```
`IllegalState`

xEdziu commented on src/main/java/pwr/isa/klama/rentalItem/RentalItemService.java:
```diff
@@ -0,0 +1,327 @@
+        }
+
+        RentalItem rentalItemToUpdate = rentalItemRepository.findById(rentalItemId)
+                .orElseThrow(() -> new IllegalStateException("Przedmiot o id " + rentalItemId + " nie istnieje"));
```
`ResourceNotFound`

xEdziu reviewed (changes requested):
Pewnie masz mnie już dość, ale to tylko drobne błędy/literówki. Tylko jedna rzecz wymaga większej uwagi - API potestowałem tyle ile dałem radę, nie znalazłem czegoś więcej do czego mógłbym się przyczepić :))

ZuberRS03 committed:
Poprawione błędy z PR.
